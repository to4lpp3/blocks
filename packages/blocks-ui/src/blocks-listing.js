/** @jsx jsx */
import React, { useMemo } from 'react'
import { Droppable, Draggable } from 'react-beautiful-dnd'
import { jsx, ThemeProvider } from 'theme-ui'

import InlineBlockRender from './inline-block-render'

export default ({ components, theme }) => {
  const list = useMemo(() => {
    return Object.keys(components).map((key, i) => (
      <Draggable key={key} draggableId={key} index={i + 1}>
        {(provided, snapshot) => {
          const Component = components[key]

          return (
            <div
              ref={provided.innerRef}
              {...provided.draggableProps}
              {...provided.dragHandleProps}
            >
              <div
                sx={{
                  m: 2,
                  border: 'thin solid #e1e6eb'
                }}
              >
                <InlineBlockRender
                  sx={{
                    '& > *': {
                      m: 0,
                      p: 0
                    }
                  }}
                  aria-label={key}
                  code={Component.usage}
                  scope={{
                    [key]: Component
                  }}
                />
              </div>
            </div>
          )
        }}
      </Draggable>
    ))
  }, [components])

  return (
    <ThemeProvider theme={theme}>
      <Droppable droppableId="components">
        {(provided, snapshot) => (
          <div {...provided.droppableProps} ref={provided.innerRef}>
            {list}
            {provided.placeholder}
          </div>
        )}
      </Droppable>
    </ThemeProvider>
  )
}
