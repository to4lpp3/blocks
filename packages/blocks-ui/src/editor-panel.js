/** @jsx jsx */
import React from 'react'
import { jsx } from 'theme-ui'
import { Label, Input } from '@theme-ui/components'

// TODO: Fix this prop drilling dance for common editor
// interactions.
export default ({
  elementData,
  handleChange,
  handlePropChange,
  handleRemove,
  handleRemoveElement,
  handleParentSelect,
  handleInsertElement,
  handleClone,
  handleTextUpdate,
  setElementId,
  blocks = [] // TODO: Fix
}) => (
  <div
    sx={{
      height: '100vh',
      overflow: 'scroll'
    }}
  >
    {elementData ? (
      <h3
        sx={{
          textTransform: 'uppercase',
          fontSize: 2,
          fontWeight: 600,
          m: 0,
          px: 3,
          py: 1,
          borderBottom: 'thin solid #e1e6eb'
        }}
      >
        {elementData.name}
        <button onClick={handleRemoveElement}>Remove</button>
        <button onClick={handleInsertElement}>Insert</button>
        <button onClick={handleClone}>Clone</button>
        {elementData.parentId && (
          <button onClick={handleParentSelect}>Parent</button>
        )}
      </h3>
    ) : (
      <ul>
        {blocks.map &&
          blocks.map(block => (
            <li key={block.id} onClick={() => setElementId(block.id)}>
              {block.name}
            </li>
          ))}
      </ul>
    )}
    <div>
      {elementData && (
        <div sx={{ px: 3 }}>
          {elementData.hasOwnProperty('text') && (
            <React.Fragment>
              <Label id="text">Text</Label>
              <Input
                sx={{
                  display: 'block',
                  width: '100%'
                }}
                onChange={handleTextUpdate}
                value={elementData.text}
                aria-labelledby="text"
              />
            </React.Fragment>
          )}
          <h3
            sx={{
              fontWeight: 'normal',
              m: 0,
              pb: 2,
              pt: 4
            }}
          >
            Props
          </h3>
          {elementData.props.hasOwnProperty('to') && (
            <React.Fragment>
              <Label id="to">To</Label>
              <Input
                sx={{
                  display: 'block',
                  width: '100%'
                }}
                onChange={handlePropChange('to')}
                value={elementData.props.to || ''}
                aria-labelledby="to"
              />
            </React.Fragment>
          )}
          <h3
            sx={{
              fontWeight: 'normal',
              mt: 4,
              mb: 0,
              pt: 4,
              borderTop: 'thin solid'
            }}
          >
            Styles
          </h3>
          {elementData.props.sx && (
            <React.Fragment>
              <Label>Padding</Label>
              <Input
                sx={{
                  display: 'block',
                  width: '100%'
                }}
                onChange={handleChange('p')}
                value={elementData.props.sx.p}
              />
              <button onClick={handleRemove('p')}>Remove</button>
            </React.Fragment>
          )}
          <React.Fragment>
            <Label>Background</Label>
            <Input
              sx={{
                display: 'block',
                width: '100%'
              }}
              onChange={handleChange('background')}
              value={
                elementData.props.sx && elementData.props.sx.backgroundColor
              }
            />
            <button onClick={handleRemove('background')}>Remove</button>
          </React.Fragment>
          <React.Fragment>
            <Label>Color</Label>
            <Input
              sx={{
                display: 'block',
                width: '100%'
              }}
              onChange={handleChange('color')}
              value={elementData.props.sx.color}
            />
            <button onClick={handleRemove('color')}>Remove</button>
          </React.Fragment>
          <h3
            sx={{
              fontWeight: 'normal',
              mt: 4,
              mb: 0,
              pt: 4,
              borderTop: 'thin solid'
            }}
          >
            Variant
          </h3>
          {elementData.props.sx && elementData.props.sx.variant && (
            <React.Fragment>
              <Label>
                {elementData.props.sx.variant.replace('styles.', '')}
              </Label>
              <Input
                sx={{
                  display: 'block',
                  width: '100%'
                }}
                onChange={handleChange('variant')}
                value={elementData.props.sx.variant}
              />
            </React.Fragment>
          )}
          <pre>{JSON.stringify(elementData, null, 2)}</pre>
        </div>
      )}
    </div>
  </div>
)
