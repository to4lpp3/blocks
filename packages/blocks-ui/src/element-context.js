import React, { useContext } from 'react'

const ElementContext = React.createContext({})

export const useElement = () => {
  const element = useContext(ElementContext)
  return element || {}
}

export default ElementContext
