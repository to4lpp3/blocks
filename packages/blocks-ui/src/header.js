/** @jsx jsx */
import { jsx } from 'theme-ui'
import { Code, Layers } from 'react-feather'

import { useEditor } from './editor-context'
import { IconButton } from './ui'

export default () => {
  const editorState = useEditor()
  console.log(editorState)

  return (
    <header
      sx={{
        display: 'flex',
        width: '100%',
        py: 2,
        px: 3,
        borderBottom: 'thin solid #e1e6eb'
      }}
    >
      <a href="/" sx={{ mr: 'auto' }}>
        <img
          alt="Blocks logo"
          src="https://user-images.githubusercontent.com/1424573/61592179-e0fda080-ab8c-11e9-9109-166cc7c86b43.png"
          sx={{
            height: 20,
            verticalAlign: 'middle'
          }}
        />
      </a>
      {editorState.mode === 'canvas' ? (
        <IconButton
          aria-label="View code"
          onClick={() => {
            editorState.update({ ...editorState, mode: 'code' })
          }}
        >
          <Code size={15} sx={{ position: 'relative', top: '1px' }} />
        </IconButton>
      ) : (
        <IconButton
          aria-label="View canvas"
          onClick={() => {
            editorState.update({ ...editorState, mode: 'canvas' })
          }}
        >
          <Layers size={15} sx={{ position: 'relative', top: '1px' }} />
        </IconButton>
      )}
    </header>
  )
}
