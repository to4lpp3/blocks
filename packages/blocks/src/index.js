// HEADERS
export { default as HeaderBasic } from './headers/basic'
export { default as HeaderLogo } from './headers/logo'

// TAGLINES
export { default as TaglineBold } from './taglines/bold'
export { default as TaglineContent } from './taglines/content'
