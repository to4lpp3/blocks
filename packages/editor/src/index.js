export { default as Editor } from './components/Editor'

export { serializer, stringifyMDX } from '@blocks/serializer'
