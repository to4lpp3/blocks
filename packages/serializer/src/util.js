export const isCloseTag = (str = '') =>
  str.startsWith('</') || str.endsWith('/>')
export const isOpenTag = (str = '') => !isCloseTag(str) && str.endsWith('>')

export const getComponentName = (str = '') => {
  const match = str.match(/^\<?([\w\.\_]+)/)
  return match && match[1]
}
