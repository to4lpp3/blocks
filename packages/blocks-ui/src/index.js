/** @jsx jsx */
import React, { useState, useEffect } from 'react'
import { DragDropContext, Droppable, Draggable } from 'react-beautiful-dnd'
import { Styled, jsx } from 'theme-ui'
import * as presets from '@theme-ui/presets'

import * as DEFAULT_BLOCKS from '@blocks/blocks/src'

import * as themeComponents from '@theme-ui/components'

import * as transforms from './transforms'
import * as queries from './queries'

import pragma from './pragma'

import Header from './header'
import Canvas from './canvas'
import Layout from './layout'
import SidePanel from './side-panel'

// blocks app theme
const appTheme = {
  ...presets.system,
  styles: {
    ...presets.system.styles,
    navlink: {
      color: 'inherit',
      textDecoration: 'none',
      fontWeight: 600
    }
  },
  buttons: {
    primary: {
      color: 'background',
      bg: 'primary',
      '&:hover': {
        bg: 'text'
      }
    },
    secondary: {
      color: 'text',
      bg: 'background',
      borderColor: 'text',
      border: 'thin solid'
    }
  }
}

const Blocks = {}
Blocks.Root = React.Fragment

const BLOCKS_Draggable = ({ active, children, ...props }) => {
  return (
    <Draggable {...props}>
      {(provided, snapshot) =>
        children(
          {
            ...provided,
            draggableProps: {
              ...provided.draggableProps,
              css: {
                boxShadow: active ? 'inset 0px 0px 0px 2px #0079FF' : undefined,
                ':hover': { boxShadow: 'inset 0px 0px 0px 2px #bbbbbb' },
                '&:focus': {
                  outline: 'none',
                  boxShadow: 'inset 0px 0px 0px 1px #4d9ef7'
                }
              }
            }
          },
          snapshot
        )
      }
    </Draggable>
  )
}

export default ({ src: initialCode, blocks: providedBlocks, onChange }) => {
  const [code, setCode] = useState(null)
  const [rawCode, setRawCode] = useState(null)
  const [transformedCode, setTransformedCode] = useState(null)
  const [elementId, setElementId] = useState(null)
  const [elementData, setElementData] = useState(null)
  const [activeTab, setActiveTab] = useState(0)
  const [themeName, setThemeName] = useState('system')

  const blocks = providedBlocks ? providedBlocks : DEFAULT_BLOCKS
  const theme = presets[themeName]

  const scope = {
    Blocks,
    Styled,
    Link: Styled.a,
    jsx: pragma(setElementId),
    BLOCKS_Droppable: Droppable,
    BLOCKS_Draggable: props => (
      <BLOCKS_Draggable active={props.draggableId === elementId} {...props} />
    ),
    BLOCKS_DraggableInner: props => <div {...props} />,
    BLOCKS_DroppableInner: props => <div {...props} />,
    BLOCKS_Text: props => <span {...props} />,
    ...themeComponents,
    ...blocks
  }

  useEffect(() => {
    const newCode = transforms.addTuid(initialCode)
    setCode(newCode)
  }, [])

  useEffect(() => {
    if (!elementId) {
      return
    }

    const newElementData = queries.getCurrentElement(code, elementId)
    setElementData(newElementData)

    if (newElementData) {
      setActiveTab(0)
    }
  }, [elementId, code])

  useEffect(() => {
    try {
      const newTransformedCode = transforms.toTransformedJSX(code)
      const newRawCode = transforms.toRawJSX(code)
      setRawCode(newRawCode)

      if (newTransformedCode) {
        setTransformedCode(newTransformedCode)
      }

      if (newRawCode) {
        onChange(newRawCode)
      }
    } catch (e) {}
  }, [code])

  if (!code || !transformedCode) {
    return null
  }

  const onDragEnd = drag => {
    if (!drag.destination || drag.destination.droppableId === 'components') {
      return
    }

    if (
      drag.destination === 'root' &&
      drag.source.index === drag.destination.index
    ) {
      return
    }

    if (drag.source.droppableId === 'components') {
      const newCode = transforms.insertJSXBlock(code, { ...drag, blocks })
      setCode(newCode)
    } else if (drag.source.droppableId.startsWith('element-')) {
      console.log(drag)
    } else {
      const newCode = transforms.reorderJSXBlocks(code, drag)
      setCode(newCode)
    }
  }

  const onBeforeDragStart = drag => {
    setElementId(drag.draggableId)
  }

  const handleRemove = key => () => {
    const sx = elementData.props.sx || {}
    delete sx[key]

    setElementData({ ...elementData, props: { ...elementData.props, sx } })

    const { code: newCode } = transforms.removeSxProp(code, { elementId, key })
    setCode(newCode)
  }

  const handleRemoveElement = () => {
    const { code: newCode } = transforms.removeElement(code, { elementId })
    setCode(newCode)
    setElementId(null)
    setElementData(null)
  }

  const handleClone = () => {
    const { code: newCode } = transforms.cloneElement(code, { elementId })
    setCode(newCode)
  }

  const handleInsertElement = () => {
    const { code: newCode } = transforms.insertElementAfter(code, { elementId })
    setCode(newCode)
  }

  const handleTextUpdate = e => {
    const text = e.target.value
    setElementData({ ...elementData, text })

    const { code: newCode } = transforms.replaceText(code, { text, elementId })
    setCode(newCode)
  }

  const handleChange = newSx => {
    const sx = elementData.props.sx || {}

    setElementData({
      ...elementData,
      props: { ...elementData.props, sx: { ...sx, ...newSx } }
    })

    const { code: newCode } = transforms.applySxProp(code, {
      elementId,
      sx: newSx
    })

    setCode(newCode)
  }

  const handleParentSelect = () => {
    if (elementData.parentId) {
      setElementId(elementData.parentId)
    }
  }

  const handlePropChange = key => e => {
    setElementData({
      ...elementData,
      props: { ...elementData.props, [key]: e.target.value }
    })

    const { code: newCode } = transforms.applyProp(code, {
      elementId,
      key,
      value: e.target.value
    })

    setCode(newCode)
  }

  console.log(elementData)

  return (
    <Layout elementData={elementData} theme={theme}>
      <Header />
      <DragDropContext
        onDragEnd={onDragEnd}
        onBeforeDragStart={onBeforeDragStart}
      >
        <div
          sx={{
            display: 'flex'
          }}
        >
          <Canvas
            code={rawCode}
            transformedCode={transformedCode}
            scope={scope}
            theme={theme}
          />
          <SidePanel
            activeTab={activeTab}
            setActiveTab={setActiveTab}
            blocks={blocks}
            theme={appTheme}
            themeName={themeName}
            elementData={elementData}
            handleChange={handleChange}
            handlePropChange={handlePropChange}
            handleRemove={handleRemove}
            handleRemoveElement={handleRemoveElement}
            handleParentSelect={handleParentSelect}
            handleInsertElement={handleInsertElement}
            handleClone={handleClone}
            handleTextUpdate={handleTextUpdate}
            setThemeName={setThemeName}
            setElementId={setElementId}
          />
        </div>
      </DragDropContext>
    </Layout>
  )
}
