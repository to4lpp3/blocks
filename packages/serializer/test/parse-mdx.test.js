import { parseMDX } from '../src'

const FIXTURE = `
# Hello, __world!__

<Block>

## Other

Stuff

</Block>
`

test('it parses a basic block', () => {
  const result = parseMDX('# Hello, world!')

  expect(result).toMatchSnapshot()
})
