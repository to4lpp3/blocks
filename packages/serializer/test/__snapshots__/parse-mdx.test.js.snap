// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it parses a basic block 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": "Hello, world!",
        },
      ],
      "depth": 1,
      "type": "heading",
    },
  ],
  "type": "root",
}
`;
